File: Contensive40Version/ResourceLibrary/vbConversion/fastIndexClass.vb
Option Explicit On
Option Strict On

Imports Contensive.BaseClasses

Namespace Contensive.VbConversion
    Public Class fastIndexClass
        '
        ' New serializable And deserialize
        '   Declare a private instance of a class that holds everything()
        '   keyPtrIndex uses the Class
        '   Call serialize On keyPtrIndex To json serialize the storage Object And Return the String
        '   Call deserialise To populate the storage Object from the argument

        ' ----- Index Type - This structure Is the basis for Element Indexing
        '       Records are read into thier data Structure, And keys(Key,ID,etc.) And pointers
        '       are put In the KeyPointerArrays.
        '           AddIndex( Key, value )
        '           'BubbleSort( Index ) - sorts the index by the key field
        '           GetIndexValue( index, Key ) - retrieves the pointer

        ' These  GUIDs provide the COM identity For this Class 
        ' And its COM interfaces. If you change them, existing 
        ' clients will no longer be able To access the Class.
        'Public Const ClassId As String = "BB8AFA32-1C0A-4CDB-BE3B-D9E6AA91A656"
        'Public Const InterfaceId As String = "353333D8-FB3B-4340-B8B6-C5547B46F5DF"
        'Public Const EventsId As String = "1407C7AD-08DF-44DB-898E-7B3CB9F86EB3"
        '
        Private Const KeyPointerArrayChunk = 1000
        '
        <Serializable()>
        Public Class storageClass
            '
            Public ArraySize As Integer
            Public ArrayCount As Integer
            Public ArrayDirty As Boolean
            Public UcaseKeyArray() As String
            Public PointerArray() As String
            Public ArrayPointer As Integer
        End Class
        '
        Private store As New storageClass
        '
        '
        '
        'Public Function exportPropertyBag() As String
        '    Dim returnBag As String = ""
        '    Try
        '        Dim json As New System.Web.Script.Serialization.JavaScriptSerializer
        '        '

        '        returnBag = json.Serialize(store)
        '        'returnBag = Newtonsoft.Json.JsonConvert.SerializeObject(store)
        '        'Catch ex As Newtonsoft.Json.JsonException
        '        '    Throw New indexException("ExportPropertyBag JSON error", ex)
        '    Catch ex As Exception
        '        Throw New indexException("ExportPropertyBag error", ex)
        '    End Try
        '    Return returnBag
        'End Function
        ''
        ''
        ''
        'Public Sub importPropertyBag(ByVal bag As String)
        '    Try
        '        Dim json As New System.Web.Script.Serialization.JavaScriptSerializer
        '        '
        '        store = json.Deserialize(Of storageClass)(bag)
        '        'store = Newtonsoft.Json.JsonConvert.DeserializeObject(Of storageClass)(bag)
        '        'Catch ex As Newtonsoft.Json.JsonException
        '        '    Throw New indexException("ImportPropertyBag JSON error", ex)
        '    Catch ex As Exception
        '        Throw New indexException("ImportPropertyBag error", ex)
        '    End Try
        'End Sub
        '
        '========================================================================
        '   Returns a pointer into the index for this Key
        '   Used only by GetIndexValue and setIndexValue
        '   Returns -1 if there is no match
        '========================================================================
        '
        Private Function GetArrayPointer(ByVal Key As String) As Integer
            Dim ArrayPointer As Integer = -1
            Try
                Dim UcaseTargetKey As String
                'Dim ElementKey As String
                Dim HighGuess As Integer
                Dim LowGuess As Integer
                Dim PointerGuess As Integer
                Dim test As String
                test = ""
                '
                If store.ArrayDirty Then
                    Call Sort()
                End If
                '
                ArrayPointer = -1
                If store.ArrayCount > 0 Then
                    UcaseTargetKey = Key.ToUpper().Replace(vbCrLf, "")
                    LowGuess = -1
                    HighGuess = store.ArrayCount - 1
                    Do While (HighGuess - LowGuess) > 1
                        ' 20150823 jk added to prevent implicit conversion
                        PointerGuess = CInt(Int((HighGuess + LowGuess) / 2))
                        'PointerGuess = (HighGuess + LowGuess) / 2
                        If UcaseTargetKey = store.UcaseKeyArray(PointerGuess) Then
                            HighGuess = PointerGuess
                            Exit Do
                        ElseIf UcaseTargetKey < store.UcaseKeyArray(PointerGuess) Then
                            HighGuess = PointerGuess
                        Else
                            LowGuess = PointerGuess
                        End If
                    Loop
                    If UcaseTargetKey = store.UcaseKeyArray(HighGuess) Then
                        ArrayPointer = HighGuess
                    End If
                End If

            Catch ex As Exception
                Throw New indexException("getArrayPointer error", ex)
            End Try
            Return ArrayPointer
        End Function
        '
        '========================================================================
        '   Returns the matching pointer from a ContentIndex
        '   Returns -1 if there is no match
        '========================================================================
        '
        Public Function GetPointer(key As String) As Integer
            Return getPtr(key)
        End Function
        Public Function getPtr(ByVal Key As String) As Integer
            Dim returnKey As Integer = -1
            Try
                Dim test As String
                Dim MatchFound As Boolean
                Dim UcaseKey As String
                test = ""
                '
                UcaseKey = Key.ToUpper().Replace(vbCrLf, "")
                'UcaseKey = genericController.vbUCase(Key)
                store.ArrayPointer = GetArrayPointer(Key)
                If store.ArrayPointer > -1 Then
                    ' Make sure this is the first match
                    MatchFound = True
                    Do While MatchFound
                        store.ArrayPointer = store.ArrayPointer - 1
                        If store.ArrayPointer < 0 Then
                            MatchFound = False
                        Else
                            MatchFound = (store.UcaseKeyArray(store.ArrayPointer) = UcaseKey)
                        End If
                    Loop
                    store.ArrayPointer = store.ArrayPointer + 1

                    returnKey = 0
                    If (IsNumeric(store.PointerArray(store.ArrayPointer))) Then
                        returnKey = CInt(store.PointerArray(store.ArrayPointer))
                    End If
                End If
            Catch ex As Exception
                Throw New indexException("GetPointer error", ex)
            End Try
            Return returnKey
        End Function
        '
        '========================================================================
        '   Add an element to an ContentIndex
        '
        '   if the entry is a duplicate, it is added anyway
        '========================================================================
        '
        Public Sub SetPointer(key As String, pointer As Integer)
            setPtr(key, pointer)
        End Sub
        Public Sub setPtr(ByVal Key As String, ByVal Pointer As Integer)
            Try
                Dim keyToSave As String
                '
                keyToSave = Key.ToUpper().Replace(vbCrLf, "")
                '
                If store.ArrayCount >= store.ArraySize Then
                    store.ArraySize = store.ArraySize + KeyPointerArrayChunk
                    ReDim Preserve store.PointerArray(store.ArraySize)
                    ReDim Preserve store.UcaseKeyArray(store.ArraySize)
                End If
                store.ArrayPointer = store.ArrayCount
                store.ArrayCount = store.ArrayCount + 1
                store.UcaseKeyArray(store.ArrayPointer) = keyToSave
                store.PointerArray(store.ArrayPointer) = CStr(Pointer)
                store.ArrayDirty = True
            Catch ex As Exception
                Throw New indexException("SetPointer error", ex)
            End Try
        End Sub
        '
        '========================================================================
        '   Returns the next matching pointer from a ContentIndex
        '   Returns -1 if there is no match
        '========================================================================
        '
        Public Function GetNextPointerMatch(key As String) As Integer
            Return getNextPtrMatch(key)
        End Function
        '
        Public Function getNextPtrMatch(ByVal Key As String) As Integer
            Dim nextPointerMatch As Integer = -1
            Try
                Dim UcaseKey As String
                '
                If store.ArrayPointer < (store.ArrayCount - 1) Then
                    store.ArrayPointer = store.ArrayPointer + 1
                    UcaseKey = Key.ToUpper()
                    If (store.UcaseKeyArray(store.ArrayPointer) = UcaseKey) Then
                        If (IsNumeric(store.PointerArray(store.ArrayPointer))) Then
                            nextPointerMatch = CInt(store.PointerArray(store.ArrayPointer))
                        End If
                        'nextPointerMatch = genericController.EncodeInteger(store.PointerArray(store.ArrayPointer))
                    Else
                        store.ArrayPointer = store.ArrayPointer - 1
                    End If
                End If
            Catch ex As Exception
                Throw New indexException("GetNextPointerMatch error", ex)
            End Try
            Return nextPointerMatch
        End Function
        '
        '========================================================================
        '   Returns the first Pointer in the current index
        '   returns empty if there are no Pointers indexed
        '========================================================================
        '
        Public Function GetFirstPointer() As Integer
            Return getFirstPtr()
        End Function
        Public Function getFirstPtr() As Integer
            Dim firstPointer As Integer = -1
            Try
                If store.ArrayDirty Then
                    Call Sort()
                End If
                '
                ' GetFirstPointer = -1
                If store.ArrayCount > 0 Then
                    store.ArrayPointer = 0
                    firstPointer = 0
                    If IsNumeric(store.PointerArray(store.ArrayPointer)) Then
                        firstPointer = CInt(store.PointerArray(store.ArrayPointer))
                    End If
                End If
                '
            Catch ex As Exception
                Throw New indexException("GetFirstPointer error", ex)
            End Try
            Return firstPointer
        End Function
        '
        '========================================================================
        '   Returns the next Pointer, past the last one returned
        '   Returns empty if the index is at the end
        '========================================================================
        '
        Public Function GetNextPointer() As Integer
            Return getNextPtr()
        End Function
        Public Function getNextPtr() As Integer
            Dim nextPointer As Integer = -1
            Try
                If store.ArrayDirty Then
                    Call Sort()
                End If
                '
                'nextPointer = -1
                If (store.ArrayPointer + 1) < store.ArrayCount Then
                    store.ArrayPointer = store.ArrayPointer + 1
                    nextPointer = 0
                    If IsNumeric(store.PointerArray(store.ArrayPointer)) Then
                        nextPointer = CInt(store.PointerArray(store.ArrayPointer))
                    End If
                End If
            Catch ex As Exception
                Throw New indexException("GetPointer error", ex)
            End Try
            Return nextPointer
        End Function
        '
        '========================================================================
        '
        '========================================================================
        '
        Private Sub BubbleSort()
            Try
                Dim TempUcaseKey As String
                Dim tempPtrString As String
                'Dim TempPointer as integer
                Dim CleanPass As Boolean
                Dim MaxPointer As Integer
                Dim SlowPointer As Integer
                Dim FastPointer As Integer
                Dim test As String
                Dim PointerDelta As Integer
                test = ""
                '
                If store.ArrayCount > 1 Then
                    PointerDelta = 1
                    MaxPointer = store.ArrayCount - 2
                    For SlowPointer = MaxPointer To 0 Step -1
                        CleanPass = True
                        For FastPointer = MaxPointer To (MaxPointer - SlowPointer) Step -1
                            If store.UcaseKeyArray(FastPointer) > store.UcaseKeyArray(FastPointer + PointerDelta) Then
                                TempUcaseKey = store.UcaseKeyArray(FastPointer + PointerDelta)
                                tempPtrString = store.PointerArray(FastPointer + PointerDelta)
                                store.UcaseKeyArray(FastPointer + PointerDelta) = store.UcaseKeyArray(FastPointer)
                                store.PointerArray(FastPointer + PointerDelta) = store.PointerArray(FastPointer)
                                store.UcaseKeyArray(FastPointer) = TempUcaseKey
                                store.PointerArray(FastPointer) = tempPtrString
                                CleanPass = False
                            End If
                        Next
                        If CleanPass Then
                            Exit For
                        End If
                    Next
                End If
                store.ArrayDirty = False
            Catch ex As Exception
                Throw New indexException("BubbleSort error", ex)
            End Try
        End Sub
        '
        '========================================================================
        '
        ' Made by Michael Ciurescu (CVMichael from vbforums.com)
        ' Original thread: http://www.vbforums.com/showthread.php?t=231925
        '
        '========================================================================
        '
        Private Sub QuickSort()
            Try
                If store.ArrayCount >= 2 Then
                    Call QuickSort_Segment(store.UcaseKeyArray, store.PointerArray, 0, store.ArrayCount - 1)
                End If
            Catch ex As Exception
                Throw New indexException("QuickSort error", ex)
            End Try
        End Sub
        '
        '
        '========================================================================
        '
        ' Made by Michael Ciurescu (CVMichael from vbforums.com)
        ' Original thread: http://www.vbforums.com/showthread.php?t=231925
        '
        '========================================================================
        '
        Private Sub QuickSort_Segment(ByVal C() As String, ByVal P() As String, ByVal First As Integer, ByVal Last As Integer)
            Try
                Dim Low As Integer, High As Integer
                Dim MidValue As String
                Dim TC As String
                Dim TP As String
                '
                Low = First
                High = Last
                MidValue = C((First + Last) \ 2)
                '
                Do
                    While C(Low) < MidValue
                        Low = Low + 1
                    End While
                    While C(High) > MidValue
                        High = High - 1
                    End While
                    If Low <= High Then
                        TC = C(Low)
                        TP = P(Low)
                        C(Low) = C(High)
                        P(Low) = P(High)
                        C(High) = TC
                        P(High) = TP
                        Low = Low + 1
                        High = High - 1
                    End If
                Loop While Low <= High
                If First < High Then
                    QuickSort_Segment(C, P, First, High)
                End If
                If Low < Last Then
                    QuickSort_Segment(C, P, Low, Last)
                End If
            Catch ex As Exception
                Throw New indexException("QuickSort_Segment error", ex)
            End Try
        End Sub
        '
        '
        '
        Private Sub Sort()
            Try
                Call QuickSort()
                store.ArrayDirty = False
            Catch ex As Exception
                Throw New indexException("Sort error", ex)
            End Try
        End Sub
    End Class
    '
    '
    '

    Public Class indexException
        Inherits System.Exception
        Implements System.Runtime.Serialization.ISerializable
        '
        'Private _message As String

        '


        Public Sub New()
            MyBase.New()
            ' Add implementation.
        End Sub

        Public Sub New(ByVal message As String)
            MyBase.New(message)
            ' Add implementation.
        End Sub

        Public Sub New(ByVal message As String, ByVal inner As Exception)
            MyBase.New(message, inner)
            ' Add implementation.
        End Sub

        ' This constructor is needed for serialization.
        Protected Sub New(ByVal info As System.Runtime.Serialization.SerializationInfo, ByVal context As System.Runtime.Serialization.StreamingContext)
            MyBase.New(info, context)
            ' Add implementation.
        End Sub
    End Class

End Namespace

File: source/ResourceLibrary/Legacy/fastIndexClass.cs

using System;
using System.Runtime.Serialization;

namespace Contensive.Addons.ResourceLibrary.Controllers {
    public class FastIndexClass {
        private const int KeyPointerArrayChunk = 1000;

        [Serializable]
        public class storageClass {
            public int ArraySize;
            public int ArrayCount;
            public bool ArrayDirty;
            public string[] UcaseKeyArray;
            public string[] PointerArray;
            public int ArrayPointer;
        }

        private storageClass store = new storageClass();

        private int GetArrayPointer(string Key) {
            int ArrayPointer = -1;
            try {
                if (store.ArrayDirty) {
                    Sort();
                }
                ArrayPointer = -1;
                if (store.ArrayCount > 0) {
                    string UcaseTargetKey = Key.ToUpper().Replace("\r\n", "");
                    int LowGuess = -1;
                    int HighGuess = store.ArrayCount - 1;
                    while ((HighGuess - LowGuess) > 1) {
                        int PointerGuess = (HighGuess + LowGuess) / 2;
                        if (UcaseTargetKey == store.UcaseKeyArray[PointerGuess]) {
                            HighGuess = PointerGuess;
                            break;
                        } else if (string.CompareOrdinal(UcaseTargetKey, store.UcaseKeyArray[PointerGuess]) < 0) {
                            HighGuess = PointerGuess;
                        } else {
                            LowGuess = PointerGuess;
                        }
                    }
                    if (UcaseTargetKey == store.UcaseKeyArray[HighGuess]) {
                        ArrayPointer = HighGuess;
                    }
                }
            } catch (Exception ex) {
                throw new indexException("getArrayPointer error", ex);
            }
            return ArrayPointer;
        }

        public int GetPointer(string key) {
            return getPtr(key);
        }

        public int getPtr(string Key) {
            int returnKey = -1;
            try {
                string UcaseKey = Key.ToUpper().Replace("\r\n", "");
                store.ArrayPointer = GetArrayPointer(Key);
                if (store.ArrayPointer > -1) {
                    bool MatchFound = true;
                    while (MatchFound) {
                        store.ArrayPointer = store.ArrayPointer - 1;
                        if (store.ArrayPointer < 0) {
                            MatchFound = false;
                        } else {
                            MatchFound = (store.UcaseKeyArray[store.ArrayPointer] == UcaseKey);
                        }
                    }
                    store.ArrayPointer = store.ArrayPointer + 1;
                    returnKey = 0;
                    if (int.TryParse(store.PointerArray[store.ArrayPointer], out _)) {
                        returnKey = int.Parse(store.PointerArray[store.ArrayPointer]);
                    }
                }
            } catch (Exception ex) {
                throw new indexException("GetPointer error", ex);
            }
            return returnKey;
        }

        public void SetPointer(string key, int pointer) {
            setPtr(key, pointer);
        }

        public void setPtr(string Key, int Pointer) {
            try {
                string keyToSave = Key.ToUpper().Replace("\r\n", "");
                if (store.ArrayCount >= store.ArraySize) {
                    store.ArraySize = store.ArraySize + KeyPointerArrayChunk;
                    Array.Resize(ref store.PointerArray, store.ArraySize + 1);
                    Array.Resize(ref store.UcaseKeyArray, store.ArraySize + 1);
                }
                store.ArrayPointer = store.ArrayCount;
                store.ArrayCount = store.ArrayCount + 1;
                store.UcaseKeyArray[store.ArrayPointer] = keyToSave;
                store.PointerArray[store.ArrayPointer] = Pointer.ToString();
                store.ArrayDirty = true;
            } catch (Exception ex) {
                throw new indexException("SetPointer error", ex);
            }
        }

        public int GetNextPointerMatch(string key) {
            return getNextPtrMatch(key);
        }

        public int getNextPtrMatch(string Key) {
            int nextPointerMatch = -1;
            try {
                if (store.ArrayPointer < (store.ArrayCount - 1)) {
                    store.ArrayPointer = store.ArrayPointer + 1;
                    string UcaseKey = Key.ToUpper();
                    if (store.UcaseKeyArray[store.ArrayPointer] == UcaseKey) {
                        if (int.TryParse(store.PointerArray[store.ArrayPointer], out _)) {
                            nextPointerMatch = int.Parse(store.PointerArray[store.ArrayPointer]);
                        }
                    } else {
                        store.ArrayPointer = store.ArrayPointer - 1;
                    }
                }
            } catch (Exception ex) {
                throw new indexException("GetNextPointerMatch error", ex);
            }
            return nextPointerMatch;
        }

        public int GetFirstPointer() {
            return getFirstPtr();
        }

        public int getFirstPtr() {
            int firstPointer = -1;
            try {
                if (store.ArrayDirty) {
                    Sort();
                }
                if (store.ArrayCount > 0) {
                    store.ArrayPointer = 0;
                    firstPointer = 0;
                    if (int.TryParse(store.PointerArray[store.ArrayPointer], out _)) {
                        firstPointer = int.Parse(store.PointerArray[store.ArrayPointer]);
                    }
                }
            } catch (Exception ex) {
                throw new indexException("GetFirstPointer error", ex);
            }
            return firstPointer;
        }

        public int GetNextPointer() {
            return getNextPtr();
        }

        public int getNextPtr() {
            int nextPointer = -1;
            try {
                if (store.ArrayDirty) {
                    Sort();
                }
                if ((store.ArrayPointer + 1) < store.ArrayCount) {
                    store.ArrayPointer = store.ArrayPointer + 1;
                    nextPointer = 0;
                    if (int.TryParse(store.PointerArray[store.ArrayPointer], out _)) {
                        nextPointer = int.Parse(store.PointerArray[store.ArrayPointer]);
                    }
                }
            } catch (Exception ex) {
                throw new indexException("GetPointer error", ex);
            }
            return nextPointer;
        }

        private void BubbleSort() {
            try {
                if (store.ArrayCount > 1) {
                    int PointerDelta = 1;
                    int MaxPointer = store.ArrayCount - 2;
                    for (int SlowPointer = MaxPointer; SlowPointer >= 0; SlowPointer--) {
                        bool CleanPass = true;
                        for (int FastPointer = MaxPointer; FastPointer >= (MaxPointer - SlowPointer); FastPointer--) {
                            if (string.CompareOrdinal(store.UcaseKeyArray[FastPointer], store.UcaseKeyArray[FastPointer + PointerDelta]) > 0) {
                                string TempUcaseKey = store.UcaseKeyArray[FastPointer + PointerDelta];
                                string tempPtrString = store.PointerArray[FastPointer + PointerDelta];
                                store.UcaseKeyArray[FastPointer + PointerDelta] = store.UcaseKeyArray[FastPointer];
                                store.PointerArray[FastPointer + PointerDelta] = store.PointerArray[FastPointer];
                                store.UcaseKeyArray[FastPointer] = TempUcaseKey;
                                store.PointerArray[FastPointer] = tempPtrString;
                                CleanPass = false;
                            }
                        }
                        if (CleanPass) {
                            break;
                        }
                    }
                }
                store.ArrayDirty = false;
            } catch (Exception ex) {
                throw new indexException("BubbleSort error", ex);
            }
        }

        private void QuickSort() {
            try {
                if (store.ArrayCount >= 2) {
                    QuickSort_Segment(store.UcaseKeyArray, store.PointerArray, 0, store.ArrayCount - 1);
                }
            } catch (Exception ex) {
                throw new indexException("QuickSort error", ex);
            }
        }

        private void QuickSort_Segment(string[] C, string[] P, int First, int Last) {
            try {
                int Low = First;
                int High = Last;
                string MidValue = C[(First + Last) / 2];
                do {
                    while (string.CompareOrdinal(C[Low], MidValue) < 0) {
                        Low = Low + 1;
                    }
                    while (string.CompareOrdinal(C[High], MidValue) > 0) {
                        High = High - 1;
                    }
                    if (Low <= High) {
                        string TC = C[Low];
                        string TP = P[Low];
                        C[Low] = C[High];
                        P[Low] = P[High];
                        C[High] = TC;
                        P[High] = TP;
                        Low = Low + 1;
                        High = High - 1;
                    }
                } while (Low <= High);
                if (First < High) {
                    QuickSort_Segment(C, P, First, High);
                }
                if (Low < Last) {
                    QuickSort_Segment(C, P, Low, Last);
                }
            } catch (Exception ex) {
                throw new indexException("QuickSort_Segment error", ex);
            }
        }

        private void Sort() {
            try {
                QuickSort();
                store.ArrayDirty = false;
            } catch (Exception ex) {
                throw new indexException("Sort error", ex);
            }
        }
    }

    public class indexException : System.Exception, System.Runtime.Serialization.ISerializable {
        public indexException() : base() { }
        public indexException(string message) : base(message) { }
        public indexException(string message, Exception inner) : base(message, inner) { }
        protected indexException(System.Runtime.Serialization.SerializationInfo info, System.Runtime.Serialization.StreamingContext context) : base(info, context) { }
    }
}
